### DISCUSSION TITLE
Query with four nested associations using LoadWith and ThenLoad

### DISCUSSION THREAD

**Discussion by dogaanismail** [linq2db/linq2db#3064] [Q&A] [answered]
Hello,

I have a question that is about using LoadWith and ThenLoad. I need to get the fourth association.

I can get the third association. For instance, **Post -> CreatedUser -> UserDetail**, but I can not the fourth association. 

I can not get **UserDetail** for the following query . **UserDetail** returns null.  **Post -> PostComment -> CreatedUser -> UserDetail** 

How can I fix my query? 

            IEnumerable<PostListDto> data = _postRepository.Table
                 .LoadWith(x => x.CreatedUser).ThenLoad(x => x.UserDetail)
                .LoadWith(x => x.PostImages).LoadWith(x => x.PostVideos)
                .LoadWith(x => x.PostComments).ThenLoad(y => y.CreatedUser).ThenLoad(mt => mt.UserDetail) //here is the fourth
                .Select(p => new PostListDto
                {
                    Id = p.Id,
                    Text = p.Text,
                    CreatedDate = p.CreatedDate,
                    ImageUrlList = p.PostImages.Count > 0 ? p.PostImages.Select(x => x.ImageUrl).ToList() : new List<string>(),
                    VideoUrl = p.PostVideos.Count == 0 ? "" : p.PostVideos.FirstOrDefault().VideoUrl,
                    CreatedByUserName = p.CreatedUser == null ? "" : p.CreatedUser.UserName,
                    CreatedByUserPhoto = p.CreatedUser == null ? "" : p.CreatedUser.UserDetail.ProfilePhotoPath,
                    PostType = p.PostType,
                    FancyboxData = $"post{p.Id}",
                    Comments = p.PostComments.ToList().Select(y => new PostCommentListDto
                    {
                        Text = y.Text,
                        CreatedDate = y.CreatedDate,
                        Id = y.Id,
                        CreatedByUserName = y.CreatedUser == null ? "" : y.CreatedUser.UserName,
                        CreatedByUserPhoto = y.CreatedUser == null ? "" : y.CreatedUser.UserDetail.ProfilePhotoPath, **//UserDetail returns null**
                        PostId = y.PostId
                    }).ToList()                  
                }).OrderByDescending(sa => sa.CreatedDate).AsEnumerable();`

``

```

### Environment details
linq2db version: *3.2.3*
Database Server: *Microsoft Sql Server*
Database Provider: *MSSQL*
Operating system: *Windows 10*
.NET Framework: *.Net 5*

**Comment by sdanyliv** [comment, marked as answer]
It is mistake thinking that LoadWith will work with custom projection:
```cs
var data = _postRepository.Table
    .Select(p => new PostListDto
    {
        Id = p.Id,
        Text = p.Text,
        CreatedDate = p.CreatedDate,
        ImageUrlList = p.PostImages.Select(x => x.ImageUrl).ToList(),
        VideoUrl = p.PostVideos.FirstOrDefault().VideoUrl ?? "",
        CreatedByUserName = p.CreatedUser.UserName ?? "",
        CreatedByUserPhoto = p.CreatedUser.UserDetail.ProfilePhotoPath ?? "",
        PostType = p.PostType,
        FancyboxData = $"post{p.Id}",
        Comments = p.PostComments.Select(y => new PostCommentListDto
        {
            Text = y.Text,
            CreatedDate = y.CreatedDate,
            Id = y.Id,
            CreatedByUserName = y.CreatedUser.UserName ?? "",
            CreatedByUserPhoto = y.CreatedUser.UserDetail.ProfilePhotoPath ?? "", 
            PostId = y.PostId
        }).ToList()                  
    })
    .OrderByDescending(sa => sa.CreatedDate)
    .AsEnumerable();
```

**Comment by dogaanismail** [comment]
Thank you very much, it works perfectly. I have one more question about LoadWith and ThenLoad extension methods. In which condition should I use them? Can you please give some examples for them? @sdanyliv

**Comment by sdanyliv** [comment]
Only if you do not have `Select` at the end of the query. Usually if you are doing projecting to DTO - you do not need `LoadWith` at all. It is true for EF also.
